### PR TITLE
Hooks feature

### PR DESCRIPTION
**Pull request by adams85**
### Describe the purpose of your pull request

* Adds `ClientReady`, `FlagEvaluated`, `ConfigChanged`, `Error` and `BeforeClientDispose` events to `ConfigCatClientsOptions`/`ConfigCatClients`.

Also improves/fixes minor issues of the previously submitted PRs:
* To be able to meet the requirement of raising the `ClientReady` event when "maxInitWaitTimeSeconds passed" in the case of auto polling, I needed to change the `ManualResetEventSlim`-based approach to initialization handling to a `CancellationTokenSource`-based one because CTS supports active and timed notifications via `RegisterCallback`+`CancelAfter`.
* Logging around value evaluation (`RolloutEvaluator`) has been also revised and cleaned up because the original implementation didn't play nicely with the `FlagEvaluated` event and mass-evaluation like `GetAllValues` and `GetAllVariationIds`.

- [x] I have covered the applied changes with automated tests.
- [x] I have executed the full automated test set against my changes.
- [x] I have validated my changes against all supported platform versions.
- [x] I have read and accepted the [contribution agreement](https://github.com/configcat/legal/blob/main/contribution-agreement.md).

### CODE REVIEW

File: src/ConfigCatClient/Hooks.cs [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+        /// <summary>
+        /// The <see cref="System.Exception"/> object related to the error (if any).
+        /// </summary>
+        public Exception Exception { get; }
+    }
+
+    /// <summary>
+    /// Defines hooks (events) for providing notifications of <see cref="ConfigCatClient"/>'s actions.
+    /// </summary>
+    public interface IProvidesHooks
```
Comment (endret) [inline src/ConfigCatClient/Hooks.cs:62]: please move into a single file

**Review by endret — APPROVED** [review]

**Review by z4kn4fein — APPROVED** [review]

File: src/ConfigCatClient/ConfigCatClient.cs [hunk 2]
```diff
@@ -240,63 +253,85 @@ public static IConfigCatClient Get(string sdkKey, Action<ConfigCatClientOptions>
...
         {
+            T value;
+            EvaluationDetails<T> evaluationDetails;
+            SettingsWithRemoteConfig settings = default;
             try
             {
-                var settings = await this.GetSettingsAsync().ConfigureAwait(false);
-                return this.configEvaluator.Evaluate(settings.Value, key, defaultValue, user ?? this.defaultUser, settings.RemoteConfig);
+                settings = await this.GetSettingsAsync().ConfigureAwait(false);
+                value = this.configEvaluator.Evaluate(settings.Value, key, defaultValue, user ?? this.defaultUser, settings.RemoteConfig, this.log, out evaluationDetails);
```
Comment (endret) [inline src/ConfigCatClient/ConfigCatClient.cs:284]: If you assume that the `this.defaultUser` field could be read/write by multiple threads, consider making a local copy of the field (if the `user` is null) to ensure the same value as the try and the catch block too.

Comment (endret) [inline src/ConfigCatClient/ConfigCatClient.cs:284]: If you assume that the `this.defaultUser` field could be read/write by multiple threads, consider making a local copy of the field (if the user is null) to ensure the same value as the try and the catch block too.

Comment (adams85) [reply src/ConfigCatClient/ConfigCatClient.cs]: Eagle eyes there. 😄 Nice catch! 👍

**Review by z4kn4fein — APPROVED** [review]

**Pull request merged**